docs/source/reference/types.rst

```diff
@@ -100,6 +100,122 @@ the beginning or the end of the index specification::
+decorator option.
+
+For instance, consider an example where the Numba jit compiled
+function applies user-specified functions as a composition to an input
+argument::
+
+    >>> @numba.njit
+    ... def composition(funcs, x):
+    ...     r = x
+    ...     for f in funcs[::-1]:
+    ...         r = f(r)
+    ...     return r
+    ...
+    >>> @numba.cfunc("double(double)")
+    ... def a(x):
+    ...     return x + 1.0
+    ...
+    >>> @numba.njit
+    ... def b(x):
+    ...     return x * x
+    ...
+    >>> composition((a, b), 0.5), 0.5 ** 2 + 1
+    (1.25, 1.25)
+    >>> composition((b, a, b, b, a), 0.5), b(a(b(b(a(0.5)))))
+    (36.75390625, 36.75390625)
+
+Here, ``cfunc`` decorated functions ``a`` and ``b`` are considered as
+first-class function objects because these are passed in to the Numba
+jit compiled function ``composition`` as arguments, that is, the
+``composition`` is jit compiled independently from its argument function
+objects (that are collected in the input argument ``funcs``).
+
+Currently, first-class function objects can be Numba ``cfunc`` compiled
+functions, ``jit`` compiled functions, and objects that implement
+Wrapper Address Protocol (WAP) with the following restrictions:
+
+========================   ==============   ============    ===========
+Context                    cfunc compiled   jit compiled    WAP objects
+========================   ==============   ============    ===========
+Can be used as arguments   yes              yes             yes
+Namespace scoping          yes              yes             no
+Can be called              yes              yes             yes
+Can be used as items       yes              yes             yes
+Automatic overload         no               yes             no
+Can be returned            yes              yes             yes
+========================   ==============   ============    ===========
+
+Wrapper Address Protocol - WAP
+++++++++++++++++++++++++++++++
+
+Wrapper Address Protocol provides an API for making any Python object
+as first-class function for Numba jit compiled functions. This assumes
+that the Python object represents a compiled function that can be
+called via its memory address (function pointer value) from Numba jit
+compiled functions. Such the so-called WAP objects must define the
+following two methods:
+
+.. method:: __wrapper_address__(self) -> int
+
+            Return the memory address of a first-class function. This
+            method is used when Numba jit compiled function tries to
+            call the given WAP instance.
+
+.. method:: signature(self) -> numba.typing.Signature
+
+            Return the signature of the given first-class
+            function. This method is used when passing in the given
+            WAP instance to Numba jit compiled function.
+
+In addition, to make WAP objects callable from Numba jit compiled
+function in object mode, the WAP object must also implement the
```

Comment:
```suggestion
functions in object mode, the WAP object must also implement the
```

Reply:
Also, "must" should perhaps be "may", the example given directly below does not implement it.

Reply:
OK. Just to be clear, the "must" applies when one needs to use the functions in *object mode*. The example works in *nopython mode*.

Reply:
Rephrased the sentence as follows
```
In addition, the WAP object may implement the ``__call__``
method. This is necessary when calling WAP objects from Numba
:term:`JIT` compiled functions in :term:`object mode`.
```

Reply:
looks good, thanks